src/prognostic_equations/implicit/auto_dense_jacobian.jl

```diff
@@ -0,0 +1,300 @@
+This is the same
+[algorithm](https://juliadiff.org/ForwardDiff.jl/stable/dev/how_it_works/#ForwardDiff's-API)
+as `ForwardDiff.jl` uses to find multidimensional derivatives, except in this
+case the values being differentiated are `FieldVector`s instead of `Vector`s.
+"""
+struct AutoDenseJacobian{S} <: JacobianAlgorithm end
+AutoDenseJacobian(max_simultaneous_derivatives = 32) =
+    AutoDenseJacobian{max_simultaneous_derivatives}()
+
+# The number of derivatives computed simultaneously by AutoDenseJacobian.
+max_simultaneous_derivatives(::AutoDenseJacobian{S}) where {S} = S
+
+function jacobian_cache(alg::AutoDenseJacobian, Y, atmos)
+    FT = eltype(Y)
+    DA = ClimaComms.array_type(Y)
+
+    FT_dual = ForwardDiff.Dual{Jacobian, FT, max_simultaneous_derivatives(alg)}
+    Y_dual = replace_parent_type(Y, FT_dual)
+    Yₜ_dual = similar(Y_dual)
+    precomputed_dual =
+        replace_parent_type(implicit_precomputed_quantities(Y, atmos), FT_dual)
+    scratch_dual = replace_parent_type(temporary_quantities(Y, atmos), FT_dual)
+
+    N = length(Fields.column(Y, 1, 1, 1))
+    n_columns = Fields.ncolumns(Y.c)
+    column_matrices = DA{FT}(undef, N, N, n_columns)
+    column_lu_factors = copy(column_matrices)
+    column_lu_vectors = DA{FT}(undef, N, n_columns)
+
+    # LinearAlgebra.I does not support broadcasting, so we need a workaround.
+    I_column_matrix = DA{FT}(undef, N, N)
+    I_column_matrix .= 0
+    I_column_matrix[diagind(I_column_matrix)] .= 1
+    I_matrix = reshape(I_column_matrix, N, N, 1)
+
+    return (;
+        Y_dual,
+        Yₜ_dual,
+        precomputed_dual,
+        scratch_dual,
+        column_matrices,
+        column_lu_factors,
+        column_lu_vectors,
+        I_matrix,
+        N_val = Val(N), # Save N as a statically inferrable parameter.
+    )
+end
+
+function update_column_matrices!(alg::AutoDenseJacobian, cache, Y, p, dtγ, t)
+    (; Y_dual, Yₜ_dual, precomputed_dual, scratch_dual, column_matrices) = cache
+    device = ClimaComms.device(Y.c)
+    column_indices = column_index_iterator(Y)
+    scalar_names = scalar_field_names(Y)
+    scalar_level_pairs = scalar_level_pair_iterator(Y)
+    batch_size = max_simultaneous_derivatives(alg)
+    batch_size_val = Val(batch_size)
+
+    p_dual_args = ntuple(Val(fieldcount(typeof(p)))) do cache_field_index
+        cache_field_name = fieldname(typeof(p), cache_field_index)
+        if cache_field_name == :precomputed
+            (; p.precomputed..., precomputed_dual...)
+        elseif cache_field_name == :scratch
+            scratch_dual
+        else
+            getfield(p, cache_field_index)
+        end
+    end
+    p_dual = AtmosCache(p_dual_args...)
+
+    batches = Iterators.partition(scalar_level_pairs, batch_size)
+    for batch in ClimaComms.threadable(device, batches)
+        # Add a unique ε to Y for each scalar level index in this batch.
+        Y_dual .= Y
+        ClimaComms.@threaded device begin
```

Comment:
This is unclear to me what this macro does exactly. Maybe a short comment would be useful.